truffle/src/main/java/org/jruby/truffle/nodes/core/InterpolatedStringNode.java

```diff
@@ -74,13 +74,14 @@ private Object concat(VirtualFrame frame, Object[] strings) {
 
         Object builder = null;
 
+        // TODO (nirvdrum 11-Jan-16) Rewrite to avoid massively unbalanced trees.
         for (Object string : strings) {
             assert RubyGuards.isRubyString(string);
 
             if (builder == null) {
                 builder = dupNode.call(frame, string, "dup", null);
             } else {
-                builder = concatNode.call(frame, builder, "concat", null, string);
+                builder = appendNode.call(frame, builder, "append", null, string);
```

Comment:
Rubinius calls this method. We had it shimmed previously (https://github.com/jruby/jruby/blob/18866ad42627363aa331cf27d53896d6d7fa359b/truffle/src/main/ruby/core/shims.rb#L68-L72). I just implemented it as a primitive as Rubinius does. The new method lives in `bootstrap/string.rb`. This was necessary because I moved a lot of `String#concat` out to Rubinus, which uses `String#append` to implement the method. Our old shim would have been cyclical.

I could indeed just call the method rather than go through the full dispatch here. I've been avoiding that generally in case we need to run through taint or frozen checks.